fix: `append_to` functionality for email accounts with IMAP folders

kamaljohnson opened the pull request:
## Problem
The `append_to` feature with `use_imap = true` in `Email Account` was broken, while receiving emails no entries of the doctype passed in the append_to field of the imap_folder table row were created. 

This was because the append_to logic for `POP` was used to handle this even if `use_imap` was set to true in the respective email_account doc

While fixing the above issue, I found another issue, i.e the field `email_sync_option` in `Email Account` was not taken into consideration for filtering out SEEN and UNSEEN emails.

### Steps to replicate
1.  Go to `Email Account` doctype list 
2. Create a new `Email Account` (configure with a test email id)
3. Select `Use IMAP`,  mark `Enable Incoming` as `true`
4. Add entries to `IMAP Folder` (eg: folder name: `Test Folder` append to: `ToDo`, make sure you set the `IMAP Folder` in your email provider also)
5. Save the doc
6. Now send a mail to the mail id used to set the `Email Account` (since the email pull does not happen immediately, you could go to `Scheduled Job Type > email_account.pull` click on `execute` this will do the pull immediately
7. Now what we expect is the `communication` for the corresponding mail to be created and the `doctype reference` for the `communication` be set to `ToDo` and also create a new `ToDo` doc (since we have set the `append_to` as `ToDo`.
8. But this does not happen, instead only the communication will be created and the `reference_doctype` will not even be set 

## Solution
Now, `append_to` doctype passed in the imap_folder will be considered while creating and linking doctype_references to the communications created.

## Limitation
Currently, emails with shared IMAP folders will not be handled, this is because, these emails share the same uid so if communication is created then while handling the other ones, no new communication will be created, rather the existing communication will be reloaded ( this logic was set before, and not introduced in this PR ). Therefore among the shared IMAP folders the one mentioned first ( order is considered from the order of the IMAP folder table in `Email Account` ) will be handled ( i.e. communications will be created according to the details mentioned by the IMAP folder )

kamaljohnson committed:
fix: auto creation of append to doctype ref while receiving mail

kamaljohnson commented:
There is one more issue that needs to be fixed, the communication timeline.

kamaljohnson commented:
There is one more issue to be fixed, the timeline emails are not shown in both communications and also the append_to doctypes refs

ankush commented:
> There is one more issue to be fixed, the timeline emails are not shown in both communications and also the append_to doctypes refs

fixed here: https://github.com/frappe/frappe/pull/15821

kamaljohnson committed:
fix: append_to parameter order in InboundMail

kamaljohnson committed:
feat: add filter for append_to wrt imap_folder

kamaljohnson committed:
refactor: remove debug logs

kamaljohnson commented:
There is another issue, it seems like when you select 'use_imap' the 'Email Sync Option' is not considered into the logic of fetch, it fetches all the emails regardless of the Option. will have to check this too.

kamaljohnson committed:
fix: email syncing with "seen" emails even if email_sync filter set to "unseen" issue

kamaljohnson committed:
Merge branch 'develop' into fix-imap-folder-append-to-not-working-issue

kamaljohnson committed:
feat: add test

kamaljohnson committed:
fix: append_to in case of not passed issue

kamaljohnson committed:
fix: reply emails get_reference_doc from communication issue

kamaljohnson committed:
fix: communication if is_first and no append_to issue

kamaljohnson committed:
feat: add tests for imap folders email syncing with multiple folders

kamaljohnson committed:
fix: remove messages-1.raw

kamaljohnson committed:
fix: sider styling issue

kamaljohnson committed:
fix: issue when test_mails are passed while testing

kamaljohnson committed:
Merge branch 'develop' into fix-imap-folder-append-to-not-working-issue

kamaljohnson committed:
fix: messages and test_mail not passed issue while running test

surajshetty3416 committed:
fix: Remove `in_test` flag inside `get_inbound_mails`

- Use mocked data instead for testing

surajshetty3416 committed:
Merge branch 'develop' into fix-imap-folder-append-to-not-working-issue

kamaljohnson committed:
fix: add back `in_test` override for `imap_folders` test

Since a lot of tests still uses the old way i.e, use `in_test` and execute custome (non-production) code to get results. Kept the same code for now else as a major test refactor needs to be done.

kamaljohnson committed:
Merge branch 'develop' into fix-imap-folder-append-to-not-working-issue

kamaljohnson committed:
fix: add in_test parameter from frappe.local.flags.in_test mock

kamaljohnson committed:
Merge branch 'develop' into fix-imap-folder-append-to-not-working-issue

kamaljohnson committed:
Merge branch 'develop' into fix-imap-folder-append-to-not-working-issue

kamaljohnson committed:
Merge branch 'develop' into fix-imap-folder-append-to-not-working-issue

kamaljohnson committed:
fix: none type issue in email test, add more coverage to `test_auto_reply`

kamaljohnson committed:
feat: add coverage for some more email tests

- test_incoming
- test_incoming_with_attach
- test_incoming_attached_email_from_outlook_plain_text_only
- test_incoming_attached_email_from_outlook_layers

kamaljohnson committed:
fix: test fail issue for some email tests

- test_threading
- test_threading_by_message_id
- test_threading_by_subject

kamaljohnson committed:
fix: use new mocked_inbound_mail in test_email

- remove in_test block from production code

surajshetty3416 committed:
test: Fix mocked method and formatting issues

kamaljohnson committed:
refactor: code cleanup
